rudder-core/src/main/scala/com/normation/rudder/reports/execution/ReportsExecutionRepository.scala

```diff
@@ -71,4 +71,61 @@ trait WoReportsExecutionRepository {
    */
   def updateExecutions(executions : Seq[AgentRun]) : Box[Seq[AgentRun]]
 
-}
\ No newline at end of file
+}
+
+/**
+ * A cached version of the service that only look in the underlying data (expected to
+ * be slow) when no cache available.
+ */
+class CachedReportsExecutionRepository(
+    readBackend : RoReportsExecutionRepository
+  , writeBackend: WoReportsExecutionRepository
+) extends RoReportsExecutionRepository with WoReportsExecutionRepository with CachedRepository {
+
+  /**
+   * We need to synchronise on cache to avoid the case:
+   * - initial state: RUNS_0 in backend
+   * - write RUNS_1 (cache => None) : ok, only write in backend
+   * [interrupt before actual write]
+   * - read: cache = None => update cache with RUNS_0
+   * - cache = RUNS_0
+   * [resume write]
+   * - write in backend RUNS_1
+   *
+   * => cache will never ses RUNS_1 before a clear.
+   */
+
+  private[this] var cache: Option[Map[NodeId, AgentRun]] = None
+
+
+  override def clearCache(): Unit = this.synchronized {
+    cache = None
+  }
+
+  override def getNodesLastRun(nodeIds: Set[NodeId]): Box[Map[NodeId, Option[AgentRun]]] = this.synchronized {
+    cache match {
+      case Some(c) => Full( nodeIds.map(x => (x, c.get(x))).toMap )
+      case None =>
+        for {
+          runs <- readBackend.getNodesLastRun(nodeIds)
+        } yield {
+          cache = Some(runs.collect{ case (nodeId, Some(run)) => (nodeId, run) }.toMap)
+          runs
+        }
+    }
+  }
+
+
+  override def updateExecutions(executions : Seq[AgentRun]) : Box[Seq[AgentRun]] = this.synchronized {
+    for {
+      runs <- writeBackend.updateExecutions(executions)
+    } yield {
+      cache match {
+        case None => //ok, does nothing
```

Comment:
Because we would miss all the node run that are not updated here and already in backend. So we must always initialise cache (from None to Some...) with a full read. 

That make me thing that it is actually NOT what I did, since I initialized the cache with a partial read only. So dumb. 

Let me correct that.